### PR TITLE
On Unix, only explicitly terminate the native process if not in background

### LINKED ISSUE #18181: macOS: opening documents by file path only is broken
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

The following is currently broken on macOS 12.6:

```powershell
'hi' > t.txt
./t.txt  # !! Nothing happens.
```

Note that `Invoke-Item ./t.txt` works.

### Expected behavior

`./t.txt` should open in the default text editor.

### Actual behavior

Nothing happens.

### Error details

_No response_

### Environment data

```powershell
PowerShell Core 7.3.0-preview.8 (also present in 7.2.6) on macOS 12.6 (M1)
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

A change was introduced in 7.2-preview.6 to explicitly terminate unix processes so they weren't orphaned, however, it killed all processes including ones that should be running in the background.  The fix is to check to make sure the process isn't in the background before performing the cleanup.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/18181

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 1]
```diff
@@ -1171,17 +1171,19 @@ private void CleanUp()
...
 #if UNIX
-                try
-                {
-                    _nativeProcess?.Kill();
-                }
-                catch
-                {
-                    // Ignore all exception since it is cleanup.
+                if (!_isRunningInBackground) {
+                    try
```
Comment (dkaszews) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:1179]: Can the two try-catches be combined? Both catch everything and do nothing

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 2]
```diff
@@ -1171,17 +1171,19 @@ private void CleanUp()
 
             try
             {
-                // on Unix, we need to kill the process to ensure it terminates as Dispose() merely
+                // on Unix, we need to kill the process (if not running in background) to ensure it terminates as Dispose() merely
```
Comment (dkaszews) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:1174]: Suggestion: add comma to make it clearer, as I initially read it as "terminates as Dispose()" and wondered how can the process terminate as a disposable.

```suggestion
                // on Unix, we need to kill the process (if not running in background) to ensure it terminates, as Dispose() merely
```

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 3]
```diff
@@ -1171,17 +1171,19 @@ private void CleanUp()
...
                 // closes the redirected streams and the processs does not exit on macOS.  However,
```
Comment (dkaszews) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:1175]: 1. The beginning of the comment mentions Unix but the end is about MacOS, can you please explain? Should it be something like "on some distributions, like MacOS"?
2. Typo and double space, I know this line was changed but it's part of the modified commit so why not fix it

```suggestion
                // closes the redirected streams and the process does not exit on macOS. However,
```

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 1 above)
Comment (dkaszews) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:1179]: Understood, the indents were a bit hard to follow

**Review by dkaszews — APPROVED** [review]

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 4]
```diff
@@ -1171,17 +1171,19 @@ private void CleanUp()
...
+                // However, on Windows, a winexe like notepad should continue running so we don't want to kill it.
 #if UNIX
-                try
-                {
-                    _nativeProcess?.Kill();
-                }
-                catch
-                {
-                    // Ignore all exception since it is cleanup.
+                if (!_isRunningInBackground) {
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:1178]:
```suggestion
                if (!_isRunningInBackground)
                {
```

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 5]
```diff
@@ -1171,17 +1171,19 @@ private void CleanUp()
 
             try
             {
-                // on Unix, we need to kill the process to ensure it terminates as Dispose() merely
-                // closes the redirected streams and the processs does not exit on macOS.  However,
-                // on Windows, a winexe like notepad should continue running so we don't want to kill it.
+                // on Unix, we need to kill the process (if not running in background) to ensure it terminates,
+                // as Dispose() merely closes the redirected streams and the process does not exit.
+                // However, on Windows, a winexe like notepad should continue running so we don't want to kill it.
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:1176]: I don't see an issue here since `Complete` is only for when the process runs in the foreground and the existing code was already there to explicitly `Dispose`

[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:1176]: I mean, when exception happens in `Prepare` or `ProcessRecord`, we may want to kill the process on Unix platforms even if it's running in background.

[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:1176]: I see, let me review that

[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:1177]: looks like this should be the case that in `Prepare` and `ProcessRecord` if something raises an exception, then we should cleanup even background processes

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 6]
```diff
@@ -1171,17 +1172,19 @@ private void CleanUp()
...
 #if UNIX
-                try
-                {
-                    _nativeProcess?.Kill();
-                }
-                catch
-                {
-                    // Ignore all exception since it is cleanup.
+                if (killBackgroundProcess || !_isRunningInBackground) {
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:1179]:
```suggestion
                if (killBackgroundProcess || !_isRunningInBackground)
                {
```

**Review by daxian-dbw — APPROVED** [review]
LGTM with one suggestion.
